test/Base_Tests/src/Data/Vector_Spec.enso

```diff
@@ -1034,6 +1030,26 @@ add_specs suite_builder =
             v = Vector.collect l .x .xs limit=30 stop_at=(_==List.Nil)
             v . should_equal [1, 2, 3]
 
+    suite_builder.group "Vector.build" group_builder->
+        group_builder.specify "Should be able to build a vector" <|
+            v = Vector.build builder->
+                builder.append 1
+                builder.append 2
+                builder.append Nothing
+                builder.append 3
+            v . should_equal [1, 2, Nothing, 3]
+
+            v2 = Vector.build builder->
+                _ = builder  # Do nothing
+            v2 . should_equal []
+
+        group_builder.specify "Should propagate dataflow errors" <|
+            v = Vector.build builder->
+                builder.append 1
+                builder.append 2
+                Error.throw (Illegal_Argument.Error "asdf")
+            v . should_fail_with (Illegal_Argument.Error "asdf")
```

Comment:
This is a great test.

Not directly related, but while we are at it we could consider solving a long outstanding problem we used to have:
```
        group_builder.specify "Should propagate the first dataflow error encountered in `append`" <|
            v = Vector.build builder->
                builder.append 1
                builder.append (Error.throw (Illegal_Argument.Error "ERROR2")
                builder.append 3
                builder.append (Error.throw (Illegal_Argument.Error "ERROR4")
                builder.append 5
            v . should_fail_with (Illegal_Argument.Error "ERROR2")
```

Unfortunately, as far as I know currently this will actually not fail but produce a vector `[1, 3, 5]`, essentially swallowing the dataflow errors.

We used to have a pattern of 'folding' the builder i.e. `append` returns a builder and this would have to be rewritten as:
```
                builder . append 1
                        . append (Error.throw (Illegal_Argument.Error "ERROR2")
                        . append 3
                        . append (Error.throw (Illegal_Argument.Error "ERROR4")
                        . append 5
```
in which case the error would be correctly propagated.

Unfortunately, the 'folding the builder' pattern did not get much traction. In many places it was just not too practical - the builder was inherently mutable object so this more 'functional' folding was often not fitting its usages.

Do you think we could amend the builder to detect that `append` got a dataflow error and avoid swallowing it?

This could be implemented by holding a `encountered_error : Ref (Any | Nothing)` inside of `Builder`, then `append` can check `item.is_error` and set this `Ref` the first time an error is encountered; then `Builder.to_vector` would first check if `encountered_error` is set and if so, propagate that error instead of building the vector.

Reply:
Very good idea. #9733